### ISSUE TITLE
Left and Right versions of the same CtElement

### ISSUE THREAD

**Issue by micheletufano** [SpoonLabs/gumtree-spoon-ast-diff#25]
Hello,

this is just a question.

Let A and B be two versions of the same file. An _insert_ action happened between the two versions. From the action I can extract the CtElement inserted. From the element, I can extract its parent method. This is the method as it appears in version B.

_Question_: how can I get the same method as it appears in version A?

**What I tried**
_From Metadata_
The following returns me the node in the version B (if is Insert action):
`CtElement element = (CtElement) action.getNode().getMetadata(SpoonGumTreeBuilder.SPOON_OBJECT);`

The following metadata:
`SpoonGumTreeBuilder.SPOON_OBJECT_DEST`
is null for Insert operation (looks like only set for Move operations).

_Inverse Diff_
I tried to perform two diffs: the A-to-B diff and the B-to-A diff. However, the insert operation (in A-to-B) become a delete operation (in B-to-A) and therefore I always get only the B version of the method.

_Manually map Method signatures_
The only way I see is to go in the file A, extract the types of CtElement I want (let's say CtMethod) and check if the signature is equal to the CtElement I got in B. This should work, except when an operation changed the signature of the method!

Do you guys have any suggestion for this? Maybe is less complicated than this and I'm just missing an API call.

p.s. Thanks for yet another great SpoonLab project! :)

**Comment by monperrus** [comment]
> An /insert/ action happened between the two versions.
> |SpoonGumTreeBuilder.SPOON_OBJECT_DEST| is null for Insert operation (looks like only set for Move
> operations). 
/This is normal and the intended behavior. Conceptually insert means that it is a new element./

> /Question/: how can I get the same method as it appears in version A?

If I understand correctly you may have found a bug, where an insert should actually be an update.

Would you have a test case?

**Comment by micheletufano** [comment]
@monperrus , I don't think is a bug, I probably was not clear enough in my explanation.
Let me rephrase:

Given an Insert action detected between the trees A and B. Is it possible to extract the **_parent_** of the inserted CtElement both in A and B? That is, in B I will get the the parent node which will contain also the inserted node. While in A, I will get the parent node without the inserted node (so the version of the parent node before the action).

p.s. Yes, the metadata totally makes sense. I was just listing what I tried while exploring the APIs.

**Comment by micheletufano** [comment]
@monperrus , your suggestion helped a lot!

Specifically, `commonAncestor()` is not what I need, since it returns the common ancestor of *all* the operations. My goal, instead, is the following: for each action, show the method (where the action was performed) before and after the change. Using commonAncestor could return the entire CtType if multiple action span across different methods. Moreover, it would return only one "version".

However, looking at the code of the method, I found that MappingStore could be really helpful for me. I got the MappingStore variable (using reflection, since it's private!), and solved with this:

**Insert Operation**
_Before the change_
```
CtElement el = (CtElement) mapping.getSrc(action.getNode().getParent()).getMetadata(SpoonGumTreeBuilder.SPOON_OBJECT);
CtMethod beforeMethod = el.getParent(CtMethod.class);
```

_After the change_
`CtMethod afterMethod = operation.getNode().getParent(CtMethod.class);`

**Delete Operation**
_Before the change_
`CtMethod beforeMethod = operation.getNode().getParent(CtMethod.class);`

_After the change_
```
CtElement el = (CtElement) mapping.getDst(action.getNode().getParent()).getMetadata(SpoonGumTreeBuilder.SPOON_OBJECT);
CtMethod afterMethod = el.getParent(CtMethod.class);
```

**Move & Update Operations**
Just use the Metadata.


@monperrus , thanks for all your help!
Please, feel free to close the issue.

**Comment by monperrus** [comment]
Excellent.

I would like to add this new method you propose. What would be the best 
API to your opinion? Would you make a pull request?

--Martin

**Comment by micheletufano** [comment]
@monperrus , I would be available to share my code.

Regarding the API, I would let you guys make the decision on the signature, which best suites the library.

On my end, when I started using this library, I was looking for some methods that were able to return me the Node in the source and destination for a given Action/Operation. Maybe this could be a good pair of methods:

`CtElement getSrcNodeForAction(Action a)`
`CtElement getDstNodeForAction(Action a)`

Or maybe the methods could return an ITree.

The CtMethod granularity is probably too specific for my task, so I would return the first mapped CtElement/ITree, instead of the parent CtMethod.

Note that, during my testing, I realized that the code I posted above could fail in some situations. For example, if the deleted node's parent is itself a deleted Node, one cannot simply get the Node from the Src or Dst tree, but rather need to traverse up the parents of the node till we find a mapped node in the Src or Dst. To this aim, the method: `Tree firstMappedSrcParent(Tree src)` in `MappingStore` could be handy.
